bcdaworker/queueing/manager/alr.go

```diff
@@ -19,64 +27,199 @@ type alrQueue struct {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// deleteAndRetry is a helper function that will delete previous attempt at
+// creating ndjson if the worker hit an error along the way. This is to avoid
+// serving incomplete or redundant data to user.
+func deleteAndRetry(l *logrus.Logger, dir string, id uint, file string) {
+	err := os.Remove(fmt.Sprintf("%s/%d/%s.ndjson", dir, id, file))
+	if err != nil {
+		l.Fatal("Could not remove alr ndjson.")
+	}
+}
+
+/******************************************************************************
+	Methods
+******************************************************************************/
+
 // startALRJob is the Job that the worker will run from the pool. This function
 // has been written here (alr.go) to separate from beneficiary FHIR workflow.
 // This job is handled by the same worker pool that works on beneficiary.
 func (q *masterQueue) startAlrJob(job *que.Job) error {
-	// Creating Context for possible cancellation
+
+	// Creating Context for possible cancellation; used by checkIfCancelled fn
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// Unmarchall JSON that contains the job details
-	// JSON selected for ALR for continuity... may change in the future
+	// Unmarshall JSON that contains the job details
 	var jobArgs models.JobAlrEnqueueArgs
 	err := json.Unmarshal(job.Args, &jobArgs)
+	// If we cannot unmarshall this, it would be very problematic.
+	if err != nil {
+		// TODO: perhaps just fail the job?
+		q.alrLog.Fatalf("Failed to unmarhall job.Args '%s' %s.",
+			job.Args, err)
+	}
+
+	// Check if this is already a failed job, and don't continue if it is
+	alrJobs, err := q.repository.GetJobByID(ctx, jobArgs.ID)
 	if err != nil {
-		q.alrLog.Warnf("Failed to unmarhall job.Args '%s' %s. Removing job...",
+		q.alrLog.Warnf("Could not get information on '%s' %s.",
 			job.Args, err)
-		// By returning nil to que-go, job is removed from queue
+		return nil
+	}
+	if alrJobs.Status == models.JobStatusCancelled {
+		q.alrLog.Warnf("ALR big job has been cancelled, worker will not tasked for %s",
+			job.Args)
 		return nil
 	}
 
-	// Check if the job was cancelled
-	go func() {
-		for {
-			select {
-			case <-time.After(15 * time.Second):
-				jobStatus, err := q.repository.GetJobByID(ctx, jobArgs.ID)
-
-				if err != nil {
-					q.alrLog.Warnf("Could not find job %d status: %s", jobArgs.ID, err)
-				}
-
-				if jobStatus.Status == models.JobStatusCancelled {
-					// cancelled context will get picked up by worker.go#writeBBDataToFile
-					cancel()
-					return
-				}
-			case <-ctx.Done():
-				return
+	// Keep track of how many times a small job has failed.
+	// If the threshold is reached, fail the job
+	alrJobTracker.Lock()
+	if _, exists := alrJobTracker.tracker[jobArgs.ID]; !exists {
+
+		alrJobTracker.tracker[jobArgs.ID] = alrJobAttempt{
+			status:         models.JobStatusInProgress,
+			attemptTracker: make(map[uint]uint),
```

Comment:
May be a good idea to pre-allocate, since small jobs may be in the 100s or 1000s.